### PR TITLE
PackedArray::fromJSON() and toJSON methods

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-2350

### CODE REVIEW

File: reference/mongodb/bson/packedarray/fromjson.xml [hunk 1]
```diff
@@ -0,0 +1,116 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- $Revision$ -->
+
+<refentry xml:id="mongodb-bson-packedarray.fromjson" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/fromjson.xml:4]: Nit: Please don't put a blank line in new files :)
```suggestion
<?xml version="1.0" encoding="utf-8"?>
<!-- $Revision$ -->
<refentry xml:id="mongodb-bson-packedarray.fromjson" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
```

File: reference/mongodb/bson/packedarray/fromjson.xml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    </listitem>
+   </varlistentry>
+  </variablelist>
+ </refsect1>
+
+ <refsect1 role="errors">
+  &reftitle.errors;
+  <simplelist>
+   &mongodb.throws.argumentparsing;
+   <member>Throws <classname>MongoDB\Driver\Exception\UnexpectedValueException</classname> if <parameter>json</parameter> is not a valid array.</member>
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/fromjson.xml:41]: Possible mistake?
```suggestion
   <member>Throws <classname>MongoDB\Driver\Exception\UnexpectedValueException</classname> if <parameter>json</parameter> is invalid JSON.</member>
```

File: reference/mongodb/bson/packedarray/fromjson.xml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  <simplelist>
+   <member><methodname>MongoDB\BSON\PackedArray::fromPHP</methodname></member>
+   <member><link xlink:href="&url.mongodb.docs.extendedjson;">MongoDB Extended JSON</link></member>
+   <member><link xlink:href="&url.mongodb.docs.bson;">BSON Types</link></member>
+  </simplelist>
+ </refsect1>
+
+</refentry>
+
+<!-- Keep this comment at the end of the file
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/fromjson.xml:97]: Ditto blank line
```suggestion
</refentry>
<!-- Keep this comment at the end of the file
```

File: reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml [hunk 4]
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- $Revision$ -->
+
+<refentry xml:id="mongodb-bson-packedarray.tocanonicalextendedjson" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml:4]: Ditto
```suggestion
<?xml version="1.0" encoding="utf-8"?>
<!-- $Revision$ -->
<refentry xml:id="mongodb-bson-packedarray.tocanonicalextendedjson" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
```

File: reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+   <member><function>MongoDB\BSON\toCanonicalExtendedJSON</function></member>
+   <member><link xlink:href="&url.mongodb.specs.extendedjson;">Extended JSON Specification</link></member>
+   <member><link xlink:href="&url.mongodb.docs.bson;">BSON Types</link></member>
+  </simplelist>
+ </refsect1>
+
+</refentry>
+
+<!-- Keep this comment at the end of the file
+Local variables:
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml:75]: Ditto
```suggestion
</refentry>
<!-- Keep this comment at the end of the file
Local variables:
```

File: reference/mongodb/bson/packedarray/torelaxedextendedjson.xml [hunk 6]
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- $Revision$ -->
+
+<refentry xml:id="mongodb-bson-packedarray.torelaxedextendedjson" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/torelaxedextendedjson.xml:4]: Ditto

File: reference/mongodb/bson/packedarray/torelaxedextendedjson.xml [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+   easily consumed by web APIs and humans.
+  </para>
+ </refsect1>
+
+ <refsect1 role="parameters">
+  &reftitle.parameters;
+  &no.function.parameters;
+ </refsect1>
+
+ <refsect1 role="examples">
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/torelaxedextendedjson.xml:30]: Missing return section

File: reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+   remain differentiated).
+  </para>
+ </refsect1>
+
+ <refsect1 role="parameters">
+  &reftitle.parameters;
+  &no.function.parameters;
+ </refsect1>
+
+ <refsect1 role="examples">
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml:31]: Missing return section

File: reference/mongodb/bson/packedarray/fromjson.xml [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    <listitem>
+     <para>
+      JSON value to be converted.
+     </para>
+    </listitem>
+   </varlistentry>
+  </variablelist>
+ </refsect1>
+
+ <refsect1 role="errors">
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/fromjson.xml:37]: Missing return section

File: reference/mongodb/bson/packedarray/torelaxedextendedjson.xml [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+   <member><methodname>MongoDB\BSON\PackedArray::toCanonicalExtendedJSON</methodname></member>
+   <member><function>MongoDB\BSON\toRelaxedExtendedJSON</function></member>
+   <member><link xlink:href="&url.mongodb.specs.extendedjson;">Extended JSON Specification</link></member>
+   <member><link xlink:href="&url.mongodb.docs.bson;">BSON Types</link></member>
+  </simplelist>
+ </refsect1>
+
+</refentry>
+
+<!-- Keep this comment at the end of the file
```
Comment (Girgias) [inline reference/mongodb/bson/packedarray/torelaxedextendedjson.xml:73]: Ditto

**Review by Girgias — CHANGES_REQUESTED** [review]
Some markup comments

File: reference/mongodb/bson/packedarray/fromjson.xml (hunk 1 above)
Comment (jmikola) [reply reference/mongodb/bson/packedarray/fromjson.xml:4]: Noted! I wasn't aware of that convention.

Out of curiosity, is it mentioned anywhere on http://doc.php.net/tutorial/?

File: reference/mongodb/bson/packedarray/fromjson.xml (hunk 2 above)
Comment (jmikola) [reply reference/mongodb/bson/packedarray/fromjson.xml:41]: This was adapted from the `Document::fromJSON()` method, which says "is not a valid document". I'll change both to say "if `$json` cannot be parsed as a document" (or "an array").

It's not just a matter of the JSON being valid. `PackedArray::fromJSON()` requires an actual array (i.e. outermost brackets) or an object with sequential numeric keys. There's some more context in https://github.com/mongodb/mongo-php-driver/pull/1703.

File: reference/mongodb/bson/packedarray/torelaxedextendedjson.xml [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+   <member><methodname>MongoDB\BSON\PackedArray::fromJSON</methodname></member>
+   <member><methodname>MongoDB\BSON\PackedArray::toCanonicalExtendedJSON</methodname></member>
+   <member><function>MongoDB\BSON\toRelaxedExtendedJSON</function></member>
+   <member><link xlink:href="&url.mongodb.specs.extendedjson;">Extended JSON Specification</link></member>
+   <member><link xlink:href="&url.mongodb.docs.bson;">BSON Types</link></member>
+  </simplelist>
+ </refsect1>
+
+</refentry>
+
```
Comment (jmikola) [inline reference/mongodb/bson/packedarray/torelaxedextendedjson.xml:72]:
```suggestion
```

File: reference/mongodb/bson/packedarray/torelaxedextendedjson.xml [hunk 12]
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- $Revision$ -->
+
```
Comment (jmikola) [inline reference/mongodb/bson/packedarray/torelaxedextendedjson.xml:3]:
```suggestion
```

File: reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml (hunk 8 above)
Comment (jmikola) [reply reference/mongodb/bson/packedarray/tocanonicalextendedjson.xml:30]: Thanks for catching this.

The return sections were also omitted from the corresponding Document docs. I'll add both.

File: reference/mongodb/bson/packedarray/fromjson.xml (hunk 1 above)
Comment (Girgias) [reply reference/mongodb/bson/packedarray/fromjson.xml]: Yeah between `<refsect1>` is OK. I don't think it is mentioned in the tutorial as that hasn't been updated in a while :/

**Review by Girgias — APPROVED** [review]
XML changes look good to me :)

**Pull request merged**